api/server/middleware.go

```diff
@@ -2,67 +2,128 @@ package server
+	// context.Context
+
+	// separating this out so we can use it and don't have to reimplement context.Context above
+	ginContext     *gin.Context
+	server         *Server
+	functionCalled bool
 }
 
-// MiddlewareFunc is a here to allow a plain function to be a middleware.
-type MiddlewareFunc func(next http.Handler) http.Handler
+// CallFunction bypasses any further gin routing and calls the function directly
+func (c *middlewareController) CallFunction(w http.ResponseWriter, r *http.Request) {
+	c.functionCalled = true
+	ctx := r.Context()
+	ctx = context.WithValue(ctx, fnext.MiddlewareControllerKey, c)
+	r = r.WithContext(ctx)
+	c.ginContext.Request = r
+	c.server.handleFunctionCall(c.ginContext)
+	c.ginContext.Abort()
+}
+func (c *middlewareController) FunctionCalled() bool {
+	return c.functionCalled
+}
 
-// Chain used to allow middlewarefuncs to be middleware.
-func (m MiddlewareFunc) Chain(next http.Handler) http.Handler {
-	return m(next)
+func (s *Server) apiMiddlewareWrapper() gin.HandlerFunc {
+	return func(c *gin.Context) {
+		s.runMiddleware(c, s.apiMiddlewares)
+	}
 }
 
-func (s *Server) middlewareWrapperFunc(ctx context.Context) gin.HandlerFunc {
+func (s *Server) rootMiddlewareWrapper() gin.HandlerFunc {
 	return func(c *gin.Context) {
-		if len(s.middlewares) > 0 {
-			defer func() {
-				//This is so that if the server errors or panics on a middleware the server will still respond and not send eof to client.
-				err := recover()
-				if err != nil {
-					common.Logger(c.Request.Context()).WithField("MiddleWarePanicRecovery:", err).Errorln("A panic occurred during middleware.")
-					handleErrorResponse(c, ErrInternalServerError)
-				}
-			}()
-			var h http.Handler
-			keepgoing := false
-			h = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-				c.Request = c.Request.WithContext(r.Context())
-				keepgoing = true
-			})
-
-			s.chainAndServe(c.Writer, c.Request, h)
-			if !keepgoing {
-				c.Abort()
-			}
+		s.runMiddleware(c, s.rootMiddlewares)
+	}
+}
+
+// This is basically a single gin middleware that runs a bunch of fn middleware.
+// The final handler will pass it back to gin for further processing.
+func (s *Server) runMiddleware(c *gin.Context, ms []fnext.Middleware) {
+	if len(ms) == 0 {
+		c.Next()
+		return
+	}
+	defer func() {
+		//This is so that if the server errors or panics on a middleware the server will still respond and not send eof to client.
+		err := recover()
+		if err != nil {
+			common.Logger(c.Request.Context()).WithField("MiddleWarePanicRecovery:", err).Errorln("A panic occurred during middleware.")
+			handleErrorResponse(c, ErrInternalServerError)
+		}
+	}()
+
+	ctx := context.WithValue(c.Request.Context(), fnext.MiddlewareControllerKey, s.newMiddlewareController(c))
+	last := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Println("final function called")
```

Comment:
see a few `Println` littered around. should we use debug logger maybe so it's there but with the flexibility to remove it? (also see that old impl used println, but easy enough to fix while we're here)